CommonTools/PileupAlgos/plugins/PuppiPhoton.cc

```diff
@@ -0,0 +1,193 @@
+  pt_                    = iConfig.getParameter<double>("pt");
+  dRMatch_               = iConfig.getParameter<std::vector<double> > ("dRMatch");
+  pdgIds_                = iConfig.getParameter<std::vector<int32_t> >("pdgids");
+  usePFRef_              = iConfig.getParameter<bool>("useRefs");
+  weight_                = iConfig.getParameter<double>("weight");
+  useValueMap_           = iConfig.getParameter<bool>("useValueMap");
+  tokenWeights_          = consumes<edm::ValueMap<float> >(iConfig.getParameter<edm::InputTag>("weightsName"));
+
+  usePhotonId_  = (iConfig.getParameter<edm::InputTag>("photonId")).label().size() == 0;
+  produces<PFOutputCollection>();
+  produces< edm::ValueMap<reco::CandidatePtr> >(); 
+}
+// ------------------------------------------------------------------------------------------
+PuppiPhoton::~PuppiPhoton(){}
+// ------------------------------------------------------------------------------------------
+void PuppiPhoton::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+
+  edm::Handle<CandidateView> hPhoProduct;
+  iEvent.getByToken(tokenPhotonCandidates_,hPhoProduct);
+  const CandidateView *phoCol = hPhoProduct.product();
+
+  edm::Handle<edm::ValueMap<bool> > photonId;
+  if(usePhotonId_) iEvent.getByToken(tokenPhotonId_,photonId);
+  int iC = -1;
+  std::vector<const reco::Candidate*> phoCands;
+  std::vector<uint16_t> phoIndx;
+
+  // Get PFCandidate Collection
+  edm::Handle<CandidateView> hPFProduct;
+  iEvent.getByToken(tokenPFCandidates_,hPFProduct);
+  const CandidateView *pfCol = hPFProduct.product();
+
+  edm::Handle<CandidateView> hPuppiProduct;
+  iEvent.getByToken(tokenPuppiCandidates_,hPuppiProduct);
+  const CandidateView *pupCol = hPuppiProduct.product();
+  for(CandidateView::const_iterator itPho = phoCol->begin(); itPho!=phoCol->end(); itPho++) {
+    iC++;
+    bool passObject = false;
+    if(itPho->isPhoton() && usePhotonId_)   passObject =  (*photonId)  [phoCol->ptrAt(iC)];
+    if(itPho->pt() < pt_) continue;
+    if(!passObject && usePhotonId_) continue;
+    if(!usePFRef_) phoCands.push_back(&(*itPho)); 
+    if(!usePFRef_) continue;
+    try{
+      const pat::Photon *pPho = dynamic_cast<const pat::Photon*>(&(*itPho));
+      if(pPho != 0) {
+	for( const edm::Ref<pat::PackedCandidateCollection> & ref : pPho->associatedPackedPFCandidates() ) {
+	  if(matchPFCandidate(&(*(pfCol->ptrAt(ref.key()))),&(*itPho))) {
+	    phoIndx.push_back(ref.key());
+	    phoCands.push_back(&(*(pfCol->ptrAt(ref.key()))));
+	  }
+	}
+	continue;
+      }
+      const pat::Electron *pElectron = dynamic_cast<const pat::Electron*>(&(*itPho));
+      if(pElectron != 0) {
+	for( const edm::Ref<pat::PackedCandidateCollection> & ref : pElectron->associatedPackedPFCandidates() ) 
+	  if(matchPFCandidate(&(*(pfCol->ptrAt(ref.key()))),&(*itPho)))  {
+	    phoIndx.push_back(ref.key());
+	    phoCands.push_back(&(*(pfCol->ptrAt(ref.key()))));
+	  }
+      }
+    } catch(...) { 
```

Comment:
(looking more carefully now)
why would there be anything to catch here?
Each case of using pPho and pElectron have a protection.

try - catch pairs are not allowed in the code where we have full control of the code that can throw inside the "try" scope.